File: contracts/ts/jobs.ts
/**
 * IJobsService manages a background job queue for async processing and retries.
 * Use for any operation >500ms, any side effect that doesn't need to block the HTTP response,
 * or any operation requiring retry logic.
 * Requires the cache module as the queue backend.
 */
export interface IJobsService {
  /**
   * Adds a job to the queue for immediate processing.
   * payload must be JSON-serializable.
   */
  enqueue(jobType: string, payload: unknown, opts?: JobOptions): Promise<JobHandle>;

  /**
   * Adds a job to the queue to be processed after the given delay in milliseconds.
   */
  enqueueIn(jobType: string, payload: unknown, delayMs: number, opts?: JobOptions): Promise<JobHandle>;

  /**
   * Adds a job to the queue to be processed at a specific time.
   */
  enqueueAt(jobType: string, payload: unknown, processAt: Date, opts?: JobOptions): Promise<JobHandle>;

  /**
   * Registers a handler function for the given job type.
   * Handlers must be idempotent — they may be called more than once for the same job.
   */
  registerHandler(jobType: string, handler: JobHandler): void;

  /**
   * Begins processing jobs from the queue.
   * Returns a promise that resolves when the processor is stopped.
   */
  start(): Promise<void>;

  /**
   * Gracefully shuts down the job processor, waiting for in-flight jobs to complete.
   */
  stop(): Promise<void>;
}

/** Returned when a job is successfully enqueued. */
export interface JobHandle {
  /** Queue-assigned unique job ID. */
  id: string;

  /** The job type that was enqueued. */
  type: string;

  /** The queue the job was placed in. */
  queue: string;
}

/**
 * A function that processes a single job.
 * payload is the deserialized job payload.
 * Return void on success. Throw an error to trigger a retry.
 */
export type JobHandler = (payload: unknown) => Promise<void>;

/** Controls how a job behaves when enqueued. */
export interface JobOptions {
  /** Which queue to place the job in. Defaults to the default queue. */
  queue?: string;

  /** Maximum number of retry attempts on failure. Defaults to global config (typically 3). */
  maxRetries?: number;

  /**
   * Prevents duplicate jobs with the same key from being enqueued.
   * If a job with the same uniqueKey already exists, the new job is dropped.
   */
  uniqueKey?: string;
}

File: contracts/go/jobs.go
package contracts

import (
	"context"
	"time"
)

// JobsService manages a background job queue for async processing and retries.
// Use for any operation >500ms, any side effect that doesn't need to block the HTTP response,
// or any operation requiring retry logic.
// Requires the cache module as the queue backend.
type JobsService interface {
	// Enqueue adds a job to the queue for immediate processing.
	// payload must be JSON-serializable.
	Enqueue(ctx context.Context, jobType string, payload any) (*JobHandle, error)

	// EnqueueIn adds a job to the queue to be processed after the given delay.
	EnqueueIn(ctx context.Context, jobType string, payload any, delay time.Duration) (*JobHandle, error)

	// EnqueueAt adds a job to the queue to be processed at a specific time.
	EnqueueAt(ctx context.Context, jobType string, payload any, processAt time.Time) (*JobHandle, error)

	// RegisterHandler registers a handler function for the given job type.
	// The handler receives the raw JSON payload; unmarshal it into the expected type.
	// Handlers must be idempotent — they may be called more than once for the same job.
	RegisterHandler(jobType string, handler JobHandler) error

	// Start begins processing jobs from the queue.
	// This is a blocking call — run it in a goroutine.
	Start(ctx context.Context) error

	// Stop gracefully shuts down the job processor, waiting for in-flight jobs to complete.
	Stop(ctx context.Context) error
}

// JobHandle is returned when a job is successfully enqueued.
type JobHandle struct {
	// ID is the queue-assigned unique job ID.
	ID string

	// Type is the job type that was enqueued.
	Type string

	// Queue is the queue the job was placed in.
	Queue string
}

// JobHandler is a function that processes a single job.
// ctx carries the request context. payload is the raw JSON bytes for the job.
// Return nil on success. Return an error to trigger a retry.
type JobHandler func(ctx context.Context, payload []byte) error

// JobOptions controls how a job behaves when enqueued.
type JobOptions struct {
	// Queue specifies which queue to place the job in.
	// If empty, uses the default queue.
	Queue string

	// MaxRetries is the maximum number of retry attempts on failure.
	// Defaults to the global configured value (typically 3).
	MaxRetries int

	// UniqueKey prevents duplicate jobs with the same key from being enqueued.
	// If a job with the same UniqueKey already exists in the queue, the new job is dropped.
	UniqueKey string
}
